### ISSUE TITLE
segfault during gblinear

### ISSUE THREAD

**Issue by pseudotensor** [dmlc/xgboost#3250]
[segfault.zip](https://github.com/dmlc/xgboost/files/1918368/segfault.zip)

Above contains self-contained script and data.

**Comment by pseudotensor** [comment]
@RAMitchell

**Comment by pseudotensor** [comment]
If I don't pass xgb_model=booster into the model.fit() then there is no segfault.

**Comment by RAMitchell** [comment]
@pseudotensor The problem is the model is configured as n_estimators=0. When we drop down from the sklearn API into the conventional XGBoost API this parameter is then input is the number of boosting rounds. The algorithm is running for 0 boosting rounds then when it tries to read the model it segfaults.

The short term solution is to set n_estimators > 0 when constructing XGBRegressor.

We could probably improve this in XGBoost by renaming this parameter to be less specific to the tree booster, as well as checking that num_rounds > 0 when train is called.

**Comment by pseudotensor** [comment]
We have also seen it segfault when one of the columns is all NaNs.  In principle we can remove them, but I'd guess the code shouldn't segfault.  gbtree would also fail with the Check failed cut->size() issue.

**Comment by pseudotensor** [comment]
This seems to be because model.best_ntree_limit is set as 0 (or stays as 0) by gblinear code.  So if we use that suggestion as n_estimators for a later gblinear call, it fails.

**Comment by pseudotensor** [comment]
But if I understand, currently n_estimators is unused by gblinear, so why should 0 lead to segfault?  Or are you saying it is used now so one can control iterations directly and via tolerance?

**Comment by RAMitchell** [comment]
N_estimators will be the number of boosting rounds.

**Comment by pseudotensor** [comment]
will be meaning is now or soon will be but not now?

**Comment by RAMitchell** [comment]
> will be meaning is now or soon will be but not now?

Is now

> And will you fill best_ntree_limit like expected?

best_ntree_limit should be populated as the best boosting iteration given the early stopping criteria.

**Comment by pseudotensor** [comment]
Interesting, the best_ntree_limit is set to 0 by gblinear.  Maybe that's because of the warm start I put in, so eventually no iterations are required?  I guess I can't rely upon best_ntree_limit then.  Thoughts?

**Comment by RAMitchell** [comment]
A slightly rewrote your script here:
```
import numpy as np
import pandas as pd
import xgboost as xgb
import pickle


def gblinear_fit(X, y, sample_weight=None, params=None, model=None, scaler=None, **kwargs):
    n_lambda = params.get('n_lambda', 1)
    min_lambda_fraction = params.get('min_lambda_fraction', 1E-6)

    col_names = tuple(X.columns)

    Xscaled = X

    Xscaled = pd.DataFrame(np.nan_to_num(Xscaled))
    Xscaled.columns = col_names

    if kwargs['eval_set'] is not None:
        for evali in range(0, len(kwargs['eval_set'])):
            Xeval = kwargs['eval_set'][evali][0]
            yeval = kwargs['eval_set'][evali][1]
            Xevalscaled = Xeval
            Xevalscaled = pd.DataFrame(np.nan_to_num(Xevalscaled))
            Xevalscaled.columns = col_names
            kwargs['eval_set'][evali] = (Xevalscaled, yeval)

    params_model = model.get_params()
    reg_lambda0 = params_model['reg_lambda']
    reg_alpha0 = params_model['reg_alpha']

    if n_lambda > 1:
        dlambda = (np.log10(min_lambda_fraction)) / (n_lambda - 1.0)
    else:
        dlambda = 0.0

    score_old = np.finfo('d').max
    for ilambda in range(0, n_lambda):
        factor = np.power(10.0, dlambda * ilambda)
        reg_lambda = reg_lambda0 * factor
        reg_alpha = reg_alpha0 * factor
        print("lambda: {}".format(reg_lambda))
        print("alpha: {}".format(reg_alpha))
        if ilambda > 0:
            booster = model.get_booster()
            booster.set_param({'reg_lambda': reg_lambda})
            model.reg_lambda = reg_lambda
            booster.set_param({'reg_alpha': reg_alpha})
            booster.set_param({'debug_verbose': 0})
            model.reg_alpha = reg_alpha
        else:
            booster = None
        assert Xscaled.shape[1] > 0, "No features in Xscaled when calling xgboost_model_fit() gblinear"
        model.fit(Xscaled, y, sample_weight=sample_weight, xgb_model=booster, eval_set=[(Xscaled, y)])
        print("Best iteration:")
        print(model.get_booster().best_iteration)


(model, X, y, sample_weight, params, scaler, kwargs) = pickle.load(
    open("f075e7c1-a0ff-4952-9a31-bee9fbbdfbe1_xgbobj.dat", "rb"))

num_rounds = 10
model = xgb.XGBRegressor(base_score=444105.22236321482, booster='gblinear',
                         debug_verbose=0,
                         learning_rate=1.0,
                         min_lambda_fraction=0.0001,
                         n_estimators=num_rounds,
                         n_jobs=4, n_lambda=20,
                         random_state=6161055, reg_alpha=2.5,
                         reg_lambda=10,
                         updater='coord_descent')

gblinear_fit(X, y, sample_weight=sample_weight, params=params, model=model, scaler=scaler, **kwargs)
```
It uses 10 iterations to fit each alpha/lambda combination. Note that you can set the learning rate to 1.0 for coordinate descent to dramatically speed up your training.

**Comment by pseudotensor** [comment]
Closing for now.
